### PR TITLE
New Advanced Configuration Options

### PR DESCRIPTION
**Pull request by nihadness**
## Current behavior before PR

1. 'Adding a new Remote From' and 'Data of existing Remotes' are in the same View.
2. It is not possible to specify object types which a user doesn't want to import.
3. While importing from different remotes, objects with the same ID are _overriden_ or not being updated.

## Desired behavior after PR is merged

1. The old (`/sync`) view will only list the Fetched Data and their Actions. New (`/sync_add`) View will be available to add a new remote.
2. A new  field will be added to 'Add' form where users can define Content Types- separated by comma, to be skipped during import.
3. New 'Prefix Logic' will be introduced and users will be able to define a prefix for each remote and use them for specific content types. It will help the system to save _similar_ objects from different Remotes separately. Prefixes will be added only to object ID's so far... If Prefix is empty, it will behave in the old, standard way.

### Screenshots
Adding New Remote Form:

image

Extended Advanced Options:
image

-----
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

**Review by juangallostra — APPROVED** [review]

File: src/senaite/sync/browser/add.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        dataform = form.get("dataform", False)
+        if not any([fetchform, dataform]):
+            return self.template()
+
+        # Handle "Fetch" action
+        if form.get("fetch", False):
+
+            url = form.get("url", "")
+            if not url.startswith("http"):
+                url = "http://{}".format(url)
```
Comment (xispa) [inline src/senaite/sync/browser/add.py]: Make "http**s**" as default's

File: src/senaite/sync/browser/add.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            domain_name = form.get("domain_name", None)
+            username = form.get("ac_name", None)
+            password = form.get("ac_password", None)
+            # check if all mandatory fields have values
+            if not all([domain_name, url, username, password]):
+                message = _("Please fill in all required fields")
+                self.add_status_message(message, "error")
+                return self.template()
+
+            import_settings = True if form.get("import_settings") == 'on' else False
```
Comment (xispa) [inline src/senaite/sync/browser/add.py]: Please, replace it as follows:
```python
import_settings = form.get('import_settings') == 'on'
```

File: src/senaite/sync/browser/add.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            if content_types:
+                content_types = [t.strip() for t in content_types.split(",")]
+                content_types = filter(lambda ct: ct in portal_types,
+                                       content_types)
+
+            if unwanted_content_types:
+                unwanted_content_types = [t.strip() for t
+                                          in unwanted_content_types.split(",")]
+                unwanted_content_types = filter(lambda ct: ct in portal_types,
+                                                unwanted_content_types)
```
Comment (xispa) [inline src/senaite/sync/browser/add.py]: I suggest to reduce a bit the complexity by adding a function to get the content types:
```python
def get_content_types(self, content_types):
    if not content_types:
        return list()
    portal_types = api.get_tool("portal_types")
    ...
    return content_types
```
So, you'll only need to do the following here:
```python
prefixable_types = self.get_content_types(form.get("prefixable_types", None))
content_types = self.get_content_types(form.get("content_types", None))
unwanted_content_types = self.get_content_types(format.get("unwanted_content_types", None))
```
Note with this change you'll be able to remove the logic for `prefixable_types` below.

File: src/senaite/sync/browser/templates/add.pt [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+              </div>
+            </div>
+          </div>
+        </div>
+
+        <input class="btn btn-success btn-sm allowMultiSubmit"
+               type="submit"
+               name="fetch"
+               i18n:attributes="value"
+               value="Save & Fetch"/>
```
Comment (xispa) [inline src/senaite/sync/browser/templates/add.pt]: Better "Save and fetch"

File: src/senaite/sync/browser/templates/sync.pt [hunk 5]
```diff
@@ -23,196 +23,40 @@
...
-                  Import users: <span tal:replace="python:view.storage[storage]['configuration'].get('import_users')"/>
+              <div
+                      tal:define="content_types python:view.storage[storage]['configuration'].get('content_types');
+                                  unwanted_content_types python:view.storage[storage]['configuration'].get('unwanted_content_types');
+                                  prefixable_types python:view.storage[storage]['configuration'].get('prefixable_types');"
+              >
+                  <b>Prefix: <span style="font-size: medium" tal:replace="python: view.storage[storage]['configuration'].get('prefix', 'Not defined')"/></b><br><br>
+                  Settings will <span tal:condition="python: not view.storage[storage]['configuration'].get('import_settings')">NOT</span>,
+                  Registry will <span tal:condition="python: not view.storage[storage]['configuration'].get('import_registry')">NOT</span>,
+                  Users will <span tal:condition="python: not view.storage[storage]['configuration'].get('import_users')">NOT</span> be imported.<br><br>
```
Comment (xispa) [inline src/senaite/sync/browser/templates/sync.pt]: Would be nice to create a function (something like `def get_storage_setting(self, parameter, section='configuration')` in the view to make the retrieval of all these settings easier. A thumb of rule is to always try to keep the tpl as much simpler as possible, with less logic as possible.

File: src/senaite/sync/browser/views.py [hunk 6]
```diff
@@ -65,13 +65,19 @@ def __call__(self):
             url = storage["credentials"]["url"]
             username = storage["credentials"]["username"]
             password = storage["credentials"]["password"]
-            content_types = storage["configuration"].get("content_types", None)
+            content_types = storage["configuration"].get("content_types", [])
+            unwanted_content_types = storage["configuration"].get("unwanted_content_types", [])
+            prefix = storage["configuration"].get("prefix", None)
+            prefixable_types = storage["configuration"].get("prefixable_types", [])
```
Comment (xispa) [inline src/senaite/sync/browser/views.py]: Same function suggested above to be called from the tpl, could be used here to get all these settings in an easier way

File: src/senaite/sync/browser/views.py [hunk 7]
```diff
@@ -100,14 +106,20 @@ def __call__(self):
...
```
Comment (xispa) [inline src/senaite/sync/browser/views.py]: ditto

File: src/senaite/sync/importstep.py [hunk 8]
```diff
@@ -346,42 +349,44 @@ def _parents_created(self, path):
...
+                # This should never happen
                 return False
-            p_local_uid = self.sh.find_unique("path", p_path).get(
-                                                    LOCAL_UID, None)
+            p_local_uid = p_row.get(LOCAL_UID, None)
             if not p_local_uid:
                 if hasattr(existing, "UID") and existing.UID():
                     p_local_uid = existing.UID()
-                    self.sh.update_by_path(p_path, local_uid=p_local_uid)
+                    self.sh.update_by_remote_path(p_path, local_uid=p_local_uid)
```
Comment (xispa) [inline src/senaite/sync/importstep.py]: Please, add some comments here to make the code more understandable for newbies

**Review by xispa — CHANGES_REQUESTED** [review]

File: src/senaite/sync/browser/add.py (hunk 1 above)
Comment (nihadness) [reply src/senaite/sync/browser/add.py]: Safety first! : )

File: src/senaite/sync/utils.py [hunk 9]
```diff
@@ -57,10 +57,9 @@ def filter_content_types(content_types):
         return ret
 
     # Get available portal types and make it all lowercase
-    portal_types = api.get_tool("portal_types")
-
+    portal_types = api.get_tool("portal_types").listContentTypes()
     ret = [t.strip() for t in content_types.split(",") if t]
-    ret = filter(lambda ct: ct.lower() in portal_types, ret)
+    ret = filter(lambda ct, types=portal_types: ct in types, ret)
```
Comment (xispa) [inline src/senaite/sync/utils.py]: Just ensure there are no duplicates
```python
ret = list(set(ret))
```

**Review by xispa — APPROVED** [review]

**Pull request merged**